### ISSUE TITLE
Endpoint contains wrong Pod IPv4 address 

### ISSUE THREAD

**Issue by rustrial** [cilium/cilium#19931]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We observed dropped traffic between Pods for which the corresponding NetworkPolicy actually should allow traffic. Upon investigation we figured out, that only one specific Pod (from a StatefulSet) as affected, and that the problem is caused by the fact that Cilium stored a wrong IPv4 address in the Pod's `CiliumEndpoint` object.

The CiliumEndpoint's `status.networking.addressing.ipv4` field contained a wrong IPv4 address (`10.8.142.14`), not matching the effective IPv4 address of the Pod (`10.8.2.32`).

We have not yet figured out whether we can reliably reproduce this, right now it looks more like this is caused by a (rare) race condition (maybe related to StatefulSet specifics). 

**Endpoint**

```yaml
apiVersion: cilium.io/v2
kind: CiliumEndpoint
metadata:
  creationTimestamp: '2022-05-23T11:39:49Z'
  generation: 2
  labels:
    app.kubernetes.io/instance: grafana-agent-metrics
    app.kubernetes.io/managed-by: grafana-agent-operator
    app.kubernetes.io/name: grafana-agent
    app.kubernetes.io/version: v0.23.0
    controller-revision-hash: grafana-agent-metrics-shard-1-788dbb9c87
    grafana-agent: grafana-agent-metrics
    nx-k8s-topology-id: 8a8f0140-b805-44db-9dfe-625c6b5df899
    operator.agent.grafana.com/name: grafana-agent-metrics
    operator.agent.grafana.com/shard: '1'
    operator.agent.grafana.com/type: metrics
    statefulset.kubernetes.io/pod-name: grafana-agent-metrics-shard-1-0
  name: grafana-agent-metrics-shard-1-0
  namespace: grafana-agent-system
  ownerReferences:
    - apiVersion: v1
      blockOwnerDeletion: true
      kind: Pod
      name: grafana-agent-metrics-shard-1-0
      uid: 296f1523-ceac-4906-b4e1-0b8f4c4d5c01
  resourceVersion: '456543044'
  uid: 1c88af6b-67c0-4043-85f7-b821165f00ab
  selfLink: >-
    /apis/cilium.io/v2/namespaces/grafana-agent-system/ciliumendpoints/grafana-agent-metrics-shard-1-0
status:
  encryption: {}
  external-identifiers:
    container-id: 1fd8732c92dbf8283a2ecd25908604c4db0420965ee7e4f0eaba1870002e017a
    k8s-namespace: grafana-agent-system
    k8s-pod-name: grafana-agent-metrics-shard-1-0
    pod-name: grafana-agent-system/grafana-agent-metrics-shard-1-0
  id: 873
  identity:
    id: 54072
    labels:
      - k8s:app.kubernetes.io/instance=grafana-agent-metrics
      - k8s:app.kubernetes.io/managed-by=grafana-agent-operator
      - k8s:app.kubernetes.io/name=grafana-agent
      - k8s:app.kubernetes.io/version=v0.23.0
      - >-
        k8s:io.cilium.k8s.namespace.labels.grafana-agent-system.tree.hnc.x-k8s.io/depth=0
      - >-
        k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=grafana-agent-system
      - >-
        k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/name=grafana-agent
      - >-
        k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/namespace=kube-system
      - k8s:io.cilium.k8s.namespace.labels.scheduling.nexxiot.com/fargate=false
      - k8s:io.cilium.k8s.policy.cluster=default
      - k8s:io.cilium.k8s.policy.serviceaccount=grafana-agent
      - k8s:io.kubernetes.pod.namespace=grafana-agent-system
  named-ports:
    - name: http-metrics
      port: 8080
      protocol: TCP
  networking:
    addressing:
      - ipv4: 10.8.142.14
    node: 10.8.3.174
  state: ready
```

**Pod**

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: grafana-agent-metrics-shard-1-0
  generateName: grafana-agent-metrics-shard-1-
  namespace: grafana-agent-system
  uid: 296f1523-ceac-4906-b4e1-0b8f4c4d5c01
  resourceVersion: '456543083'
  creationTimestamp: '2022-05-23T11:39:48Z'
  labels:
    app.kubernetes.io/instance: grafana-agent-metrics
    app.kubernetes.io/managed-by: grafana-agent-operator
    app.kubernetes.io/name: grafana-agent
    app.kubernetes.io/version: v0.23.0
    controller-revision-hash: grafana-agent-metrics-shard-1-788dbb9c87
    grafana-agent: grafana-agent-metrics
    nx-k8s-topology-id: 8a8f0140-b805-44db-9dfe-625c6b5df899
    operator.agent.grafana.com/name: grafana-agent-metrics
    operator.agent.grafana.com/shard: '1'
    operator.agent.grafana.com/type: metrics
    statefulset.kubernetes.io/pod-name: grafana-agent-metrics-shard-1-0
  annotations:
    kubectl.kubernetes.io/default-container: grafana-agent
    kubernetes.io/psp: k8s.privileged-host
  ownerReferences:
    - apiVersion: apps/v1
      kind: StatefulSet
      name: grafana-agent-metrics-shard-1
      uid: 116802d1-f8a2-43d1-863e-6de8e5ca8150
      controller: true
      blockOwnerDeletion: true
  selfLink: /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-1-0
status:
  phase: Running
  conditions:
    - type: Initialized
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:49Z'
    - type: Ready
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:54Z'
    - type: ContainersReady
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:54Z'
    - type: PodScheduled
      status: 'True'
      lastProbeTime: null
      lastTransitionTime: '2022-05-23T11:39:48Z'
  hostIP: 10.8.4.230
  podIP: 10.8.2.32
  podIPs:
    - ip: 10.8.2.32
  startTime: '2022-05-23T11:39:49Z'
  containerStatuses:
    - name: config-reloader
      state:
        running:
          startedAt: '2022-05-23T11:39:52Z'
      lastState: {}
      ready: true
      restartCount: 0
      image: quay.io/prometheus-operator/prometheus-config-reloader:v0.47.0
      imageID: >-
        quay.io/prometheus-operator/prometheus-config-reloader@sha256:0029252e7cf8cf38fc58795928d4e1c746b9e609432a2ee5417a9cab4633b864
      containerID: >-
        containerd://ade201bd0787e0e4ae9aeff887a091fad689eeec70406c65918e63908eb6a328
      started: true
    - name: grafana-agent
      state:
        running:
          startedAt: '2022-05-23T11:39:52Z'
      lastState: {}
      ready: true
      restartCount: 0
      image: docker.io/grafana/agent:v0.23.0
      imageID: >-
        docker.io/grafana/agent@sha256:a0beeaa6642c69efa472d509be2e2cf97dcb1c7e74047cca59a9452bf068f763
      containerID: >-
        containerd://8535bdd6e890f8028dc90eb68dab2ab317ee8e615b90322b81d6ce8e125e1438
      started: true
  qosClass: Burstable
spec:
  ...
```

### Cilium Version

1.11.3

### Kernel Version

5.4.188-104.359.amzn2.x86_64

### Kubernetes Version

1.21 (v1.21.5-eks-9017834)

### Sysdump

**Note**: this is a ZIP compressed BZIP2 Tar archive (had to work the system, BZIP2 to reduce size and ZIP to make Github accept the file format).

[cilium-sysdump-20220524-105632.tar.bz2.zip](https://github.com/cilium/cilium/files/8761164/cilium-sysdump-20220524-105632.tar.bz2.zip)

### Relevant log output

```shell
> hubble  observe -f --verdict DROPPED
May 24 08:49:02.049: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:02.049: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:03.077: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:03.077: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:05.093: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
May 24 08:49:05.093: 10.8.2.32:45754 <> kube-system/kube-state-metrics-2:8081 Policy denied DROPPED (TCP Flags: SYN)
```

### Anything else?

**NetworkPolicy**

Just for completeness, the corresponding `NetworkPolicy` object:

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kube-state-metrics
  namespace: kube-system
  labels:
    kustomize.toolkit.fluxcd.io/name: kube-state-metrics
    kustomize.toolkit.fluxcd.io/namespace: kube-system
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/name: kube-state-metrics
  ingress:
    - ports:
        - protocol: TCP
          port: 8080
        - protocol: TCP
          port: 8081
      from:
        - podSelector:
            matchLabels:
              app.kubernetes.io/name: grafana-agent
          namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: grafana-agent-system
    - ports:
        - protocol: TCP
          port: 8080
        - protocol: TCP
          port: 8081
      from:
        - podSelector: {}
  egress:
    - ports:
        - protocol: TCP
          port: 443
        - protocol: TCP
          port: 6443
  policyTypes:
    - Ingress
    - Egress
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aanm** [comment]
This is an occurrence of https://github.com/cilium/cilium/issues/14184

**Comment by aanm** [comment]
@rustrial can you replicate the issue?

**Comment by rustrial** [comment]
@aanm No (as mentioned above) we cannot reproduce this (yet). Nevertheless, this should not happen and maybe reconciliation loop must be fixed to check regularly whether the `CiliumEndpoint` still matches the Pod.

**Comment by timbuchwaldt** [comment]
From what it looks like we saw this on one of our clusters, too. Same behaviour - STS was rolled, suddenly a new pod has a mismatching IP in the CEP object.
Possibly interesting: We set `terminationGracePeriodSeconds: 0` in this specifc case - this could mean the pod is already starting again when Cilium gets the info the old one is being terminated.

**Comment by timbuchwaldt** [comment]
We had this happen yet again. We have switched the 0s terminationGracePeriod to 1s, hoping this resolves the issue, but ultimately we would need to have a working 0s variant. @aanm can I assist you with any more info?

**Comment by alex-berger** [comment]
We have observed this bug several times now (last time today), and it is really annoying that we cannot reproduce it. However, it's always happening with *Pod*s created for *StatefulSet*s. We observed that the following fields of the affected *CiliumEndpoint* contained wrong IPv4 addresses:
- `status.networking.addressing.ipv4`
- `status.networking.node`

Manually updating those two fields with the correct *Pod* and *Node* IPv4 address resolves the problem. However, such manual intervention is of course tedious and the underlying bug should be fixed.

**Comment by rustrial** [comment]
@aanm Obviously, this must be kind of a race condition or logic error in the endpoint manager. Is there any chance to raise the priority of this bug, such that it gets hunted down and fixed soonish?

Also any pointer to the relevant code is welcome and I will try to review it myself, though I doubt that I will understand it :-).

**Comment by aanm** [comment]
@rustrial sorry, without a reproducible it is difficult to find what is the bug. I tried to reproduce the bug but I couldn't. The source code referent to this functionality is [here](https://github.com/cilium/cilium/blob/96c75e41311daef08eaf38ce599794caaa327740/pkg/k8s/watchers/endpointsynchronizer.go#L79-L296)

**Comment by timbuchwaldt** [comment]
I have been working on making a proof of concept to reproduce this: https://github.com/timbuchwaldt/cilium-19931-poc

I have seen it to generate unstable differences where CEPs and Pods went out of sync and at least two where they became stable and stayed like this. 

You have to apply the `sts.yaml` to create a 2-pod statefulset running nginx, alter the `scheduler.sh` to select only scheduleable nodes (our workers have this brawn-label), run this, then start `run.sh` to start re-starting the pods. Upon success the `run.sh` script aborts, showing the diff between pod IPs and CEP IPs. Further validation is needed to see if this stays stable, I recommend running both  `kubectl get pods -o  wide -w` as well as `kubectl get cep -o  wide -w` to see the changes occuring.

We are still working on making this reproduce the problem cleaner, as currently it's unclear if this happes when being scheduled to the same nodes, different nodes, if labels have any effect and so on. I'll keep this updated in case we get any closer to the actual problem.

**Comment by timbuchwaldt** [comment]
And another one, now even with timestamps:

```
run.sh:

[2022-08-19 13:18:53] Restarting
[2022-08-19 13:18:53] statefulset.apps/web restarted
[2022-08-19 13:18:53] Waiting for 1 pods to be ready...
[2022-08-19 13:19:00] Waiting for partitioned roll out to finish: 1 out of 2 new pods have been updated...
[2022-08-19 13:19:00] Waiting for 1 pods to be ready...
[2022-08-19 13:19:06] partitioned roll out complete: 2 new pods have been updated...
[2022-08-19 13:19:06] Diffing
[2022-08-19 13:19:08] 10.233.68.163
[2022-08-19 13:19:08] 10.233.69.151
[2022-08-19 13:19:08] sleeping
[2022-08-19 13:19:10]
[2022-08-19 13:19:10]
[2022-08-19 13:19:10]
[2022-08-19 13:19:10] Diffing
[2022-08-19 13:19:11] Restarting
[2022-08-19 13:19:12] statefulset.apps/web restarted
[2022-08-19 13:19:12] Waiting for partitioned roll out to finish: 0 out of 2 new pods have been updated...
[2022-08-19 13:19:12] Waiting for 1 pods to be ready...
[2022-08-19 13:19:19] Waiting for partitioned roll out to finish: 1 out of 2 new pods have been updated...
[2022-08-19 13:19:19] Waiting for 1 pods to be ready...
[2022-08-19 13:19:24] partitioned roll out complete: 2 new pods have been updated...
[2022-08-19 13:19:24] Diffing
[2022-08-19 13:19:26] 1c1
[2022-08-19 13:19:26] < 10.233.69.144 10.233.69.151
[2022-08-19 13:19:26] ---
[2022-08-19 13:19:26] > 10.233.68.23 10.233.69.144



Cilium Endpoints:

[2022-08-19 13:19:00] web-1   103           36259                                                                        ready                  10.233.68.163
[2022-08-19 13:19:00] web-0   701           2120                                                                         waiting-for-identity   10.233.79.115
[2022-08-19 13:19:01] web-0   544           2120                                                                         ready                  10.233.69.151
[2022-08-19 13:19:02] web-1   103           36259                                                                        ready                  10.233.68.163
[2022-08-19 13:19:03] web-1   103           36259                                                                        ready                  10.233.68.163
[2022-08-19 13:19:04] web-0   544           2120                                                                         ready                  10.233.69.151
[2022-08-19 13:19:05] web-0   544           2120                                                                         ready                  10.233.69.151
[2022-08-19 13:19:06] web-1   103           36259                                                                        ready                  10.233.68.163
[2022-08-19 13:19:07] web-1   103           36259                                                                        waiting-for-identity   10.233.68.163
[2022-08-19 13:19:07] web-1   103           36259                                                                        waiting-for-identity   10.233.68.163
[2022-08-19 13:19:08] web-0   544           2120                                                                         ready                  10.233.69.151
[2022-08-19 13:19:09] web-0   544           2120                                                                         ready                  10.233.69.151
[2022-08-19 13:19:10] web-1   103           36259                                                                        waiting-for-identity   10.233.68.163
[2022-08-19 13:19:11] web-1   103           36259                                                                        waiting-for-identity   10.233.68.163
[2022-08-19 13:19:12] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:12] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:12] web-1   103           36259                                                                        waiting-for-identity   10.233.68.163
[2022-08-19 13:19:12] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:14] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:14] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:15] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:16] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:17] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:18] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:19] web-0   544           2120                                                                         waiting-for-identity   10.233.69.151
[2022-08-19 13:19:20] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:21] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:21] web-0   3610          2120                                                                         ready                  10.233.68.23
[2022-08-19 13:19:21] web-0   3610          2120                                                                         ready                  10.233.68.23
[2022-08-19 13:19:22] web-0   3610          2120                                                                         ready                  10.233.68.23
[2022-08-19 13:19:23] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:24] web-1   474           36259                                                                        ready                  10.233.69.144
[2022-08-19 13:19:24] web-1   474           36259                                                                        waiting-for-identity   10.233.69.144
[2022-08-19 13:19:24] web-0   3610          2120                                                                         ready                  10.233.68.23
[2022-08-19 13:19:24] web-0   544           54653                                                                        ready                  10.233.69.151
[2022-08-19 13:19:25] web-0   544           54653                                                                        ready                  10.233.69.151
[2022-08-19 13:19:26] web-1   474           36259                                                                        waiting-for-identity   10.233.69.144
[2022-08-19 13:19:33] web-1   474           31291                                                                        ready                  10.233.69.144


Pods:

[2022-08-19 13:19:00] web-0   1/1     Terminating         0          20s   10.233.79.115   k8s-worker04.corp   <none>           <none>
[2022-08-19 13:19:00] web-1   1/1     Running             0          7s    10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:00] web-0   1/1     Terminating         0          20s   10.233.79.115   k8s-worker04.corp   <none>           <none>
[2022-08-19 13:19:00] web-0   0/1     Pending             0          0s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:00] web-0   0/1     Pending             0          0s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:01] web-0   0/1     Pending             0          1s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:01] web-0   0/1     ContainerCreating   0          1s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:01] web-0   0/1     ContainerCreating   0          1s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:02] web-1   1/1     Running             0          9s    10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:03] web-1   1/1     Running             0          10s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:04] web-0   0/1     ContainerCreating   0          4s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:05] web-0   0/1     ContainerCreating   0          5s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:05] web-0   0/1     Running             0          5s    10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:06] web-1   1/1     Running             0          13s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:06] web-0   1/1     Running             0          6s    10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:07] web-1   1/1     Running             0          14s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:08] web-0   1/1     Running             0          8s    10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:09] web-0   1/1     Running             0          9s    10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:10] web-1   1/1     Running             0          17s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:11] web-1   1/1     Running             0          18s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:12] web-0   1/1     Running             0          12s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:12] web-1   1/1     Terminating         0          19s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:12] web-1   1/1     Terminating         0          19s   10.233.68.163   k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:12] web-1   0/1     Pending             0          0s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:12] web-0   1/1     Running             0          12s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:13] web-1   0/1     Pending             0          1s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:13] web-1   0/1     Pending             0          1s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:13] web-1   0/1     ContainerCreating   0          1s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:14] web-1   0/1     ContainerCreating   0          2s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:14] web-0   1/1     Running             0          14s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:15] web-0   1/1     Running             0          15s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:16] web-1   0/1     ContainerCreating   0          4s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:17] web-1   0/1     ContainerCreating   0          5s    <none>          k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:17] web-1   0/1     Running             0          5s    10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:18] web-0   1/1     Running             0          18s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:18] web-1   1/1     Running             0          6s    10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:19] web-0   1/1     Terminating         0          19s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:19] web-0   1/1     Terminating         0          19s   10.233.69.151   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:19] web-0   0/1     Pending             0          0s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:19] web-0   0/1     Pending             0          0s    <none>          <none>                                   <none>           <none>
[2022-08-19 13:19:20] web-1   1/1     Running             0          8s    10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:21] web-0   0/1     Pending             0          2s    <none>          k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:21] web-0   0/1     ContainerCreating   0          2s    <none>          k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:21] web-1   1/1     Running             0          9s    10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:21] web-0   0/1     ContainerCreating   0          2s    <none>          k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:22] web-0   0/1     ContainerCreating   0          3s    <none>          k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:23] web-1   1/1     Running             0          11s   10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:23] web-0   0/1     Running             0          4s    10.233.68.23    k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:24] web-1   1/1     Running             0          12s   10.233.69.144   k8s-worker03.corp   <none>           <none>
[2022-08-19 13:19:24] web-0   0/1     Running             0          5s    10.233.68.23    k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:24] web-0   1/1     Running             0          5s    10.233.68.23    k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:25] web-0   1/1     Running             0          6s    10.233.68.23    k8s-worker02.corp   <none>           <none>
[2022-08-19 13:19:26] web-1   1/1     Running             0          14s   10.233.69.144   k8s-worker03.corp   <none>           <none>
```

Let me know if you need more details. Generally about the cluster: It's made up of 128 Thread workers that run a very mixed workload, from quickly spawning CI jobs to hundreds of tiny apps and long running big applications, there is a constant load on it -> It's conceivable Cilium sees many events/s happening.

**Comment by mitchese** [comment]
Running @timbuchwaldt script on my small test cluster, it almost immediately gets into an inconsistent state where the CEP is missing entirely.

```
$ kubectl get pods -o wide                                                                                                                                                                                                                                                                                                                                                 
NAME    READY   STATUS    RESTARTS   AGE   IP              NODE          NOMINATED NODE   READINESS GATES
web-0   1/1     Running   0          11m   10.233.70.142   k8s-ceph-v1   <none>           <none>
web-1   1/1     Running   0          11m   10.233.71.149   k8s-ceph-v2   <none>           <none>

$ kubectl get cep -o wide
NAME    ENDPOINT ID   IDENTITY ID   INGRESS ENFORCEMENT   EGRESS ENFORCEMENT   VISIBILITY POLICY   ENDPOINT STATE   IPV4            IPV6
web-1   1162          43161                                                                        ready            10.233.71.149   
```
Here are the logs from cilium running on k8s-ceph-v1 which appear to first create the endpoint then erroneously delete it
```
...
level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
level=info msg="Processing API request with rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=4 rateLimiterSkipped=true subsys=rate uuid=ccc0ddd6-48df-4c46-8e72-2f7a4c48ff1f
level=info msg="API request released by rate limiter" maxWaitDuration=15s name=endpoint-create parallelRequests=4 rateLimiterSkipped=true subsys=rate uuid=ccc0ddd6-48df-4c46-8e72-2f7a4c48ff1f waitDurationTotal=0s
level=info msg="Create endpoint request" addressing="&{10.233.70.142 35410376-8f70-4191-b514-c52a72e1ca41  }" containerID=af5e340d9601539e81ee9ba5d78bbcacab296721b35e994be813559bc4e3e36f datapathConfiguration="<nil>" interface=lxc5fe2e6c749da k8sPodName=test-cilium-breakage/web-0 labels="[]" subsys=daemon sync-build=true
level=info msg="New endpoint" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=1889 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (blocking)" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=1889 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=false,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Reusing existing global key" key="k8s:app=nginx;k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage;k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=default;k8s:io.kubernetes.pod.namespace=test-cilium-breakage;k8s:selected=false;k8s:statefulset.kubernetes.io/pod-name=web-0;" subsys=allocator
level=info msg="Identity of endpoint changed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=1889 identity=29328 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=false,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ oldIdentity="no identity" subsys=endpoint
level=info msg="Waiting for endpoint to be generated" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=1889 identity=29328 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (non-blocking)" containerID= datapathPolicyRevision=0 desiredPolicyRevision=6 endpointID=1889 identity=29328 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=true,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Reusing existing global key" key="k8s:app=nginx;k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage;k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=default;k8s:io.kubernetes.pod.namespace=test-cilium-breakage;k8s:selected=true;k8s:statefulset.kubernetes.io/pod-name=web-0;" subsys=allocator
level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=6 endpointID=1889 identity=29328 ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Successful endpoint creation" containerID= datapathPolicyRevision=6 desiredPolicyRevision=6 endpointID=1889 identity=29328 ipv4= ipv6= k8sPodName=/ subsys=daemon
level=info msg="API call has been processed" name=endpoint-create processingDuration=555.811481ms subsys=rate totalDuration=557.219735ms uuid=ccc0ddd6-48df-4c46-8e72-2f7a4c48ff1f waitDurationTotal=0s
level=info msg="Resolving identity labels (non-blocking)" containerID= datapathPolicyRevision=6 desiredPolicyRevision=6 endpointID=1889 identity=29328 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=false,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (non-blocking)" containerID= datapathPolicyRevision=6 desiredPolicyRevision=6 endpointID=1889 identity=29328 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=true,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Resolving identity labels (non-blocking)" containerID= datapathPolicyRevision=6 desiredPolicyRevision=6 endpointID=1889 identity=29328 identityLabels="k8s:app=nginx,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=default,k8s:io.kubernetes.pod.namespace=test-cilium-breakage,k8s:selected=false,k8s:statefulset.kubernetes.io/pod-name=web-0" ipv4= ipv6= k8sPodName=/ subsys=endpoint
level=info msg="Releasing key" key="[k8s:app=nginx k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=test-cilium-breakage k8s:selected=true k8s:statefulset.kubernetes.io/pod-name=web-0]" subsys=allocator
level=info msg="Released last local use of key, invoking global release" key="[k8s:app=nginx k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=test-cilium-breakage k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=test-cilium-breakage k8s:selected=true k8s:statefulset.kubernetes.io/pod-name=web-0]" subsys=kvstorebackend
```
This is with k8s v1.23.1 and cilium 1.11.0 27e0848

**Comment by rustrial** [comment]
Today, we observed a case where a DaemonSet was restarted and one of its `Pods` ended-up without a corresponding `CiliumEndpoint` object. This can also been seen from the following excerpt from the Kubernetes logs, the `CiliumEndpoint` is deleted and never recreated again.

---------------------------------------------------------------------------------------------------------------------------------------------------------------
|       @timestamp        |  verb  |                                             requestURI                                             | responseStatus.code |
|-------------------------|--------|----------------------------------------------------------------------------------------------------|---------------------|
| 2022-09-05 14:09:12.339 | patch  | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0/status                | 200                 |
| 2022-09-05 14:09:12.593 | delete | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0                       | 200                 |
| 2022-09-05 14:09:12.598 | create | /api/v1/namespaces/grafana-agent-system/pods                                                       | 201                 |
| 2022-09-05 14:09:12.598 | create | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0/binding               | 201                 |
| 2022-09-05 14:09:12.844 | patch  | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0/status                | 200                 |
| 2022-09-05 14:09:13.342 | patch  | /apis/cilium.io/v2/namespaces/grafana-agent-system/ciliumendpoints/grafana-agent-metrics-shard-2-0 | 200                 |
| 2022-09-05 14:09:13.597 | delete | /apis/cilium.io/v2/namespaces/grafana-agent-system/ciliumendpoints/grafana-agent-metrics-shard-2-0 | 200                 |
| 2022-09-05 14:09:17.096 | patch  | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0/status                | 200                 |
| 2022-09-05 14:09:17.346 | patch  | /api/v1/namespaces/grafana-agent-system/pods/grafana-agent-metrics-shard-2-0/status                | 200                 |
--------------------------------------------------------------------------------------------------------------------------------------------------------------- 

The corresponding log lines from the Cilium Agent running on the corresponding node of the Pod:

```
2022-09-05 14:09:13	level=info msg="Create endpoint request" addressing="&{10.8.24.238 2a9a028c-554d-4b67-b81c-9a685ba47905  }" containerID=c144c3273ae6d0e0bd4938de670733de0d5af6cb6b34887bee3bc5bae6194893 datapathConfiguration="&{false false true false true 0xc003c5f0a0}" interface=lxc47dd9c769833 k8sPodName=grafana-agent-system/grafana-agent-metrics-shard-2-0 labels="[]" subsys=daemon sync-build=true
2022-09-05 14:09:13	level=info msg="New endpoint" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3114 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-09-05 14:09:13	level=info msg="Resolving identity labels (blocking)" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3114 identityLabels="k8s:app.kubernetes.io/instance=grafana-agent-metrics,k8s:app.kubernetes.io/managed-by=grafana-agent-operator,k8s:app.kubernetes.io/name=grafana-agent,k8s:app.kubernetes.io/version=v0.23.0,k8s:io.cilium.k8s.namespace.labels.grafana-agent-system.tree.hnc.x-k8s.io/depth=0,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=grafana-agent-system,k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/name=grafana-agent,k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/namespace=kube-system,k8s:io.cilium.k8s.namespace.labels.scheduling.nexxiot.com/fargate=false,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=grafana-agent,k8s:io.kubernetes.pod.namespace=grafana-agent-system" ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-09-05 14:09:13	level=info msg="Identity of endpoint changed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3114 identity=54072 identityLabels="k8s:app.kubernetes.io/instance=grafana-agent-metrics,k8s:app.kubernetes.io/managed-by=grafana-agent-operator,k8s:app.kubernetes.io/name=grafana-agent,k8s:app.kubernetes.io/version=v0.23.0,k8s:io.cilium.k8s.namespace.labels.grafana-agent-system.tree.hnc.x-k8s.io/depth=0,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=grafana-agent-system,k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/name=grafana-agent,k8s:io.cilium.k8s.namespace.labels.kustomize.toolkit.fluxcd.io/namespace=kube-system,k8s:io.cilium.k8s.namespace.labels.scheduling.nexxiot.com/fargate=false,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=grafana-agent,k8s:io.kubernetes.pod.namespace=grafana-agent-system" ipv4= ipv6= k8sPodName=/ oldIdentity="no identity" subsys=endpoint	
2022-09-05 14:09:13	level=info msg="Waiting for endpoint to be generated" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=3114 identity=54072 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-09-05 14:09:15	level=info msg="Rewrote endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=36 endpointID=3114 identity=54072 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-09-05 14:09:15	level=info msg="Successful endpoint creation" containerID= datapathPolicyRevision=36 desiredPolicyRevision=36 endpointID=3114 identity=54072 ipv4= ipv6= k8sPodName=/ subsys=daemon
```

**Comment by rustrial** [comment]
Maybe, Cilium Agent does not compare the UID in the `ownerReferences` field of the `CiliumEndpoint` object with the one on UID of the Pod. 

```yaml
...
ownerReferences:
    - apiVersion: v1
      kind: Pod
      name: grafana-agent-metrics-shard-2-0
      uid: 05ecd3ee-6c4f-4f58-9090-64ec546401f8
```

I guess it should compare the UIDs (of the Pod and the `metadata.ownerReferences` of the CiliumEndpoint) and retry creating the CiliumEndpoint unless the UIDs match.

**Comment by stephanroslen** [comment]
We observed the same a few days ago - having a `Pod` having no `CiliumEndpoint` - observed in a `StatefulSet` though that also regularly has the issue of non matching IPs in `Pod` and `CEP`.

**Comment by thejosephstevens** [comment]
In case it's useful, we saw the same thing happen in one of our clusters where a Daemonset pod ended up with a different IP than the CEP (in this case it was a node-local-dns pod). I'm not exactly sure how to reproduce, but I did notice that the node-local-dns pod was much older than the cilium agent pod (about 12 hours).

```
❯ cilium status
    /¯¯\
 /¯¯\__/¯¯\    Cilium:         OK
 \__/¯¯\__/    Operator:       OK
 /¯¯\__/¯¯\    Hubble:         OK
 \__/¯¯\__/    ClusterMesh:    disabled
    \__/

Deployment        hubble-relay       Desired: 1, Ready: 1/1, Available: 1/1
DaemonSet         cilium             Desired: 4, Ready: 4/4, Available: 4/4
Deployment        cilium-operator    Desired: 2, Ready: 2/2, Available: 2/2
Deployment        hubble-ui          Desired: 1, Ready: 1/1, Available: 1/1
Containers:       cilium             Running: 4
                  hubble-relay       Running: 1
                  cilium-operator
                  hubble-ui          Running: 1
Cluster Pods:     42/46 managed by Cilium
Image versions    cilium          quay.io/ascendio/cilium:v1.12.3: 4
                  hubble-relay    quay.io/ascendio/hubble-relay:v1.12.3: 1
                  hubble-ui       quay.io/ascendio/hubble-ui:v0.9.2: 1
                  hubble-ui       quay.io/ascendio/hubble-ui-backend:v0.9.2: 1
```

In Hubble, the traffic to the dns pod is labeled as going to an unmanaged pod.
![image](https://user-images.githubusercontent.com/5632093/204645440-d9e4d1e7-1bc5-4204-8ed4-b20e5be6ddd1.png)

The pod is here:
```
node-local-dns-r9szs                                             1/1     Running     0          2d16h   10.0.9.3        gke-bigque-bigque-4c62e518-bf5s   <none>           <none>
```

The CEP is here:
```
apiVersion: cilium.io/v2
kind: CiliumEndpoint
metadata:
  creationTimestamp: "2022-11-27T02:13:07Z"
  generation: 1
  labels:
    controller-revision-hash: 767d44798c
    k8s-app: node-local-dns
    pod-template-generation: "3"
  name: node-local-dns-r9szs
  namespace: kube-system
  ownerReferences:
  - apiVersion: v1
    kind: Pod
    name: node-local-dns-r9szs
    uid: 8b533783-8917-4410-be95-96d53e2e6087
  resourceVersion: "75660381"
  uid: fe8b90f9-78a1-4601-a50e-086e2c7e093c
status:
  encryption:
    key: 1
  external-identifiers:
    container-id: bbc2dbbc6e61bbde37edcca330eabfb2e7b6318423d73062a2ffbac298b68025
    k8s-namespace: kube-system
    k8s-pod-name: node-local-dns-r9szs
    pod-name: kube-system/node-local-dns-r9szs
  id: 1585
  identity:
    id: 16589
    labels:
    - k8s:io.cilium.k8s.namespace.labels.admission.gatekeeper.sh/ignore=true
    - k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=kube-system
    - k8s:io.cilium.k8s.namespace.labels.trusted-traffic=true
    - k8s:io.kubernetes.pod.namespace=kube-system
    - k8s:k8s-app=node-local-dns
  named-ports:
  - name: dns
    port: 53
    protocol: UDP
  - name: dns-tcp
    port: 53
    protocol: TCP
  - name: metrics
    port: 9253
    protocol: TCP
  networking:
    addressing:
    - ipv4: 10.0.9.232
    node: 192.168.4.9
  policy:
    egress:
      enforcing: false
      state: <status disabled>
    ingress:
      enforcing: false
      state: <status disabled>
  state: ready
  visibility-policy-status: <status disabled>
```

**Comment by tommyp1ckles** [comment]
@thejosephstevens Sorry got a bit confused in the other thread, was looking at the wrong comment here.

The existing PR is to specifically fix race conditions concerning pods of the same name, i.e. StatefulSets. I'm not sure how the DaemonSet pods got into that state.

Was there any Cilium Pod restarts during that time?

**Comment by thejosephstevens** [comment]
@tommyp1ckles Apologies, should've clarified before, should I open a separate issue for the daemonset IP inconsistencies I'm seeing if this github issue is focused on a race condition regarding Statefulsets? We're getting these pretty frequently now. It seems to only be happening with node local dns pods on GKE (or at least that's the only time we notice it since it impacts DNS resolution on the node, as the result at policy application time is evaluated as requests being sent to an unmanaged pod).

**Comment by tommyp1ckles** [comment]
@thejosephstevens Sorry, missed your previous message.

Just to be sure I'm correctly understanding, looking through your examples, you posted a CiliumEndpoint for the dns pod, which is advertising an IP of 10.0.9.3, same as the Pod. Is this correct and did this become different afterwards?

> We're getting these pretty frequently now. It seems to only be happening with node local dns pods on GKE (or at least that's the only time we notice it since it impacts DNS resolution on the node, as the result at policy application time is evaluated as requests being sent to an unmanaged pod).

Seems unlikely it's the upgrade in that case, however is this issue still propping up only after Cilium agent restarts?

>  should I open a separate issue for the daemonset IP inconsistencies I'm seeing if this github issue is focused on a race condition regarding Statefulsets? 

Yeah I think it might be worth starting a new issue, I don't think this has the same cause as the StatefulSet one.
